src/audio/pipeline/pipeline-stream.c

```diff
@@ -257,9 +257,54 @@ int pipeline_trigger(struct pipeline *p, struct comp_dev *host, int cmd)
 		/* Execute immediately */
 		ret = pipeline_trigger_run(p, host, cmd);
 		return ret == PPL_STATUS_PATH_STOP ? 0 : ret;
+	case COMP_TRIGGER_XRUN:
+		/*
+		 * XRUN can happen on a pipeline, not directly attached to the host,
+		 * find the original one
+		 */
+		do {
+			/* Check the opposite direction */
+			int dir = host->direction == PPL_DIR_DOWNSTREAM ? PPL_DIR_UPSTREAM :
+				PPL_DIR_DOWNSTREAM;
+			struct list_item *buffer_list = comp_buffer_list(host, dir);
+			struct list_item *clist;
+			bool found = false;
+
+			if (list_is_empty(buffer_list))
+				/* Reached the original host */
+				break;
+
+			list_for_item(clist, buffer_list) {
+				struct comp_buffer *buffer = buffer_from_list(clist,
+								struct comp_buffer, dir);
+				struct comp_dev *buffer_comp = buffer_get_comp(buffer, dir);
+
+				switch (buffer_comp->pipeline->status) {
+				case COMP_STATE_ACTIVE:
+				case COMP_STATE_PREPARE:
+					found = true;
+					break;
+				}
+
+				if (found) {
+					host = host->direction == PPL_DIR_DOWNSTREAM ?
+						buffer_comp->pipeline->source_comp :
+						buffer_comp->pipeline->sink_comp;
+					break;
+				}
+			}
+
+			if (!found) {
+				/* No active pipeline found! Should never occur. */
+				pipe_err(p, "No active pipeline found to continue pipeline %u!",
```

Comment:
error message is not logical, "No active pipeline found to continue pipeline processing" maybe?

Reply:
well, firstly "this error should never happen" (TM), if anyone ever sees it, there's a problem. We could put an `assert()` there but I don't think we do that for this kind of errors. And no, it isn't processing, it's rather linking. I'll adjust